File: src/store/record.ts
/**
 * The two line shapes the log can hold, and the only parser for them.
 *
 * The log is append-only, so a supersession cannot edit the line it closes —
 * it is a later line that shadows it. Keeping the close as its own record is
 * what lets a reader reconstruct both the old value and the instant it stopped
 * being true; an implementation that rewrote the original line would destroy
 * the history that snapshot reads and the change feed are computed from.
 *
 * Parsing rejects a line rather than the file: a process killed mid-write
 * leaves a truncated tail, and one unreadable line must never cost the store.
 */
import type { Fact, FactStatus } from "./types.js";

/** A newly asserted fact. `transient` marks TTL ephemera a sweeper may drop. */
export interface FactRecord {
	k: "fact";
	tx: number;
	transient?: boolean;
	fact: Fact;
}

/** Closes an earlier fact: status becomes superseded, validUntil is stamped. */
export interface CloseRecord {
	k: "close";
	tx: number;
	id: string;
	validUntil: string;
}

export type StoreRecord = FactRecord | CloseRecord;

export function encode(records: StoreRecord[]): string {
	return records.map((r) => `${JSON.stringify(r)}\n`).join("");
}

/** null for any line that is not a whole, well-formed record. */
export function parseRecord(line: string): StoreRecord | null {
	let raw: unknown;
	try {
		raw = JSON.parse(line);
	} catch {
		return null;
	}
	if (typeof raw !== "object" || raw === null) return null;
	const r = raw as {
		k?: unknown;
		tx?: unknown;
		transient?: unknown;
		fact?: unknown;
		id?: unknown;
		validUntil?: unknown;
	};
	const tx = typeof r.tx === "number" ? r.tx : 0;
	if (r.k === "close") {
		if (typeof r.id !== "string" || typeof r.validUntil !== "string") return null;
		return { k: "close", tx, id: r.id, validUntil: r.validUntil };
	}
	if (r.k === "fact" && isFact(r.fact)) {
		return { k: "fact", tx, transient: r.transient === true, fact: r.fact };
	}
	return null;
}

const STATUSES: FactStatus[] = ["active", "candidate", "superseded"];

function isFact(v: unknown): v is Fact {
	if (typeof v !== "object" || v === null) return false;
	const f = v as Partial<Fact>;
	const strings = [f.id, f.subject, f.predicate, f.object, f.validFrom];
	if (strings.some((s) => typeof s !== "string")) return false;
	if (typeof f.confidence !== "number") return false;
	if (!STATUSES.includes(f.status as FactStatus)) return false;
	return f.validUntil === null || typeof f.validUntil === "string";
}

File: crates/flusk-core/src/store/record.rs
//! The two line shapes the log can hold, and the only parser for them —
//! src/features/facts/record.ts, byte for byte on the encode side.
//!
//! The log is append-only, so a supersession cannot edit the line it closes;
//! it is a later line that shadows it. Parsing rejects a line rather than the
//! file: a process killed mid-write leaves a truncated tail, and one
//! unreadable line must never cost the store.
//!
//! Encoding writes the record envelope in the reference's literal key order
//! (k, tx, transient, fact / k, tx, id, validUntil) with `transient` always
//! present — parseRecord normalizes it to a boolean, so a sweep rewrite
//! stamps "transient":false even onto lines that originally omitted it.

use super::jval::{write_js_string, JVal};
use serde_json::Number;

pub enum StoreRecord {
    Fact {
        tx: Number,
        transient: bool,
        fact: JVal,
    },
    Close {
        tx: Number,
        id: String,
        valid_until: String,
    },
}

impl StoreRecord {
    pub fn tx(&self) -> f64 {
        match self {
            StoreRecord::Fact { tx, .. } | StoreRecord::Close { tx, .. } => {
                tx.as_f64().unwrap_or(0.0)
            }
        }
    }

    /// The fact this record is about: its own id, or the id a close targets.
    pub fn subject_id(&self) -> &str {
        match self {
            StoreRecord::Fact { fact, .. } => fact.get_str("id").unwrap_or(""),
            StoreRecord::Close { id, .. } => id,
        }
    }
}

pub fn encode(records: &[StoreRecord]) -> String {
    let mut out = String::new();
    for r in records {
        match r {
            StoreRecord::Fact { tx, transient, fact } => {
                out.push_str("{\"k\":\"fact\",\"tx\":");
                out.push_str(&tx.to_string());
                out.push_str(",\"transient\":");
                out.push_str(if *transient { "true" } else { "false" });
                out.push_str(",\"fact\":");
                out.push_str(&fact.to_json());
            }
            StoreRecord::Close { tx, id, valid_until } => {
                out.push_str("{\"k\":\"close\",\"tx\":");
                out.push_str(&tx.to_string());
                out.push_str(",\"id\":");
                write_js_string(id, &mut out);
                out.push_str(",\"validUntil\":");
                write_js_string(valid_until, &mut out);
            }
        }
        out.push_str("}\n");
    }
    out
}

/// None for any line that is not a whole, well-formed record.
pub fn parse_record(line: &str) -> Option<StoreRecord> {
    let raw: JVal = serde_json::from_str(line).ok()?;
    let tx = match raw.get("tx") {
        Some(JVal::Num(n)) => n.clone(),
        _ => Number::from(0),
    };
    if raw.get_str("k") == Some("close") {
        let (id, valid_until) = (raw.get_str("id")?, raw.get_str("validUntil")?);
        return Some(StoreRecord::Close {
            tx,
            id: id.to_string(),
            valid_until: valid_until.to_string(),
        });
    }
    if raw.get_str("k") == Some("fact") {
        let fact = raw.get("fact")?;
        if is_fact(fact) {
            return Some(StoreRecord::Fact {
                tx,
                transient: raw.get("transient") == Some(&JVal::Bool(true)),
                fact: fact.clone(),
            });
        }
    }
    None
}

const STATUSES: [&str; 3] = ["active", "candidate", "superseded"];

fn is_fact(v: &JVal) -> bool {
    for key in ["id", "subject", "predicate", "object", "validFrom"] {
        if v.get_str(key).is_none() {
            return false;
        }
    }
    if !matches!(v.get("confidence"), Some(JVal::Num(_))) {
        return false;
    }
    let Some(status) = v.get_str("status") else {
        return false;
    };
    if !STATUSES.contains(&status) {
        return false;
    }
    matches!(v.get("validUntil"), Some(JVal::Null) | Some(JVal::Str(_)))
}
